src/Data/Text/Utf8.hs

```diff
@@ -148,91 +154,42 @@ unpackUtf8 (Text u8data offset length) =
-decodeUtf8 cus = error $ "Invalid UTF-8 input sequence at " ++ show (take 4 cus)
-
--- TODO: Slow placeholder implementation until we can use text-2.0
-pack :: String -> Text
-pack = go . stringToByteArray
-  where
-    go !arr = Text arr 0 $ sizeofByteArray arr
-
--- TODO: Slow placeholder implementation until we can use text-2.0
-replicate :: Int -> Text -> Text
-replicate n = pack . Prelude.concat . Prelude.replicate n . unpack
-
 -- | Convert a 'Text' value into a 'T.Text' value.
 toUtf16Text :: Text -> T.Text
 toUtf16Text (Text u8data off len) =
   T.unfoldr go 0
   where
+    go :: CodeUnitIndex -> Maybe (Char, CodeUnitIndex)
     go i
-      | i >= len = Nothing
+      | i >= CodeUnitIndex len = Nothing
       | otherwise =
         let
-          (codeUnits, codePoint) = unsafeIndexCodePoint' u8data (CodeUnitIndex $ off + i)
+          (codeUnits, codePoint) = unsafeIndexCodePoint' u8data $ CodeUnitIndex off + i
         in
           Just (codePoint, i + codeUnits)
 
--- TODO: Slow placeholder implementation until we can use text-2.0
-isInfixOf :: Text -> Text -> Bool
-isInfixOf needle haystack = T.isInfixOf (toUtf16Text needle) (toUtf16Text haystack)
-
--- TODO: Naive string search, replace once we have text-2.0
--- We can't use Data.Text.Internal.Search here because that one reports UTF-16 code unit indices.
--- Complexity: O(n*m)
-indices :: Text -> Text -> [Int]
-indices needle haystack
-  | needleLen == 0 = []
-  | otherwise = go 0 0
-  where
-    CodeUnitIndex needleLen = lengthUtf8 needle
-    CodeUnitIndex haystackLen = lengthUtf8 haystack
-
-    go startIdx needleIdx
-      -- needle is longer than remaining haystack
-      | startIdx + needleLen > haystackLen = []
-      -- whole needle matched
-      | needleIdx >= needleLen = startIdx : go (startIdx + needleLen) 0
-      -- charachter mismatch
-      | needleCp /= haystackCp = go (startIdx + 1) 0
-      -- advance
-      | otherwise = go startIdx $ needleIdx + codeUnits
-      where
-        (codeUnits, needleCp) = unsafeIndexCodePoint needle $ CodeUnitIndex needleIdx
-        (_, haystackCp) = unsafeIndexCodePoint haystack $ CodeUnitIndex $ startIdx + needleIdx
+-- | Lower-case the ASCII code points A-Z and leave the rest of ASCII intact.
+{-# INLINE toLowerAscii #-}
+toLowerAscii :: Char -> Char
+toLowerAscii cp
+  | Char.isAsciiUpper cp = Char.chr (Char.ord cp + 0x20)
+  | otherwise = cp
 
-dropWhile :: (Char -> Bool) -> Text -> Text
-dropWhile predicate text =
-  let
-    len = codeUnitIndex (lengthUtf8 text)
-    go i
-      | i >= len = i
-      | otherwise =
-        let
-          (codeUnits, codePoint) = unsafeIndexCodePoint text $ CodeUnitIndex i
-        in
-          if predicate codePoint then
-            go $ i + codeUnits
-          else
-            i
+-- TODO: Slow placeholder implementation until we can use text-2.0
+{-# INLINE lowerUtf8 #-}
+lowerUtf8 :: Text -> Text
+lowerUtf8 = pack . map Char.toLower . unpack
```

Comment:
We can make it a little less slow with little effort by using `lowerCodePoint` here.

Reply:
Addressed in a4c0f15.